File: 695-max-area-of-island/max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])

        max_area=0

        def dfs(i,j, area):
            if i<0 or i>=m or j<0 or j>=n or grid[i][j]!=1:
                return area
            else:
                grid[i][j] = 0
                area += 1
                area = dfs(i,j+1,area)
                area = dfs(1+i,j,area)
                area = dfs(i,j-1,area)
                area = dfs(i-1,j,area)
            return area

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    area = dfs(i,j,0)
                    max_area = max(max_area, area)
        return max_area

File: 695-max-area-of-island/max-area-of-island.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int maxArea = 0;
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    int area = dfs(grid, i, j, m, n, 0);
                    maxArea = Math.max(maxArea, area);
                }
            }
        }
        return maxArea;
    }
    
    private int dfs(int[][] grid, int i, int j, int m, int n, int area) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != 1) {
            return area;
        }
        grid[i][j] = 0;
        area += 1;
        area = dfs(grid, i, j + 1, m, n, area);
        area = dfs(grid, i + 1, j, m, n, area);
        area = dfs(grid, i, j - 1, m, n, area);
        area = dfs(grid, i - 1, j, m, n, area);
        return area;
    }
}

